python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -63,3 +66,215 @@ def get_instance_manager_state(
+        request_id: str = "",
+    ) -> Instance:
+        instance = Instance()
+        instance.version = 0  # it will be populated by the underlying storage.
+        instance.instance_id = instance_id
+        instance.instance_type = instance_type
+        instance.launch_request_id = request_id
+        instance.total_resources.update(resources)
+        instance.status = Instance.QUEUED
+        InstanceUtil._record_status_transition(
+            instance, Instance.QUEUED, "created from InstanceUtil"
+        )
+        return instance
+
+    @staticmethod
+    def is_cloud_instance_allocated(instance: Instance) -> bool:
+        """
+        Returns True if the instance is allocated by the cloud provider, i.e.
+        there's already a cloud node for the instance.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in {
+            Instance.ALLOCATED,
+            Instance.RAY_INSTALLING,
+            Instance.RAY_RUNNING,
+            Instance.RAY_STOPPING,
+            Instance.RAY_STOPPED,
+            Instance.STOPPING,
+            Instance.RAY_INSTALL_FAILED,
+        }
+
+    @staticmethod
+    def is_ray_pending(instance: Instance) -> bool:
+        """
+        Returns True if the instance is pending ray installation or ray startup,
+        regardless of whether the instance is allocated by the cloud provider.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in [
+            Instance.UNKNOWN,
+            Instance.QUEUED,
+            Instance.REQUESTED,
+            Instance.ALLOCATED,
+            Instance.RAY_INSTALLING,
+            Instance.ALLOCATION_FAILED,
+            Instance.RAY_INSTALL_FAILED,
+        ]
+
+    @staticmethod
+    def set_status(
+        instance: Instance,
+        new_instance_status: Instance.InstanceStatus,
+        details: str = "",
+    ):
+        """Transitions the instance to the new state.
+
+        This should be called by the reconciler and the instance manager.
+
+        Args:
+            instance: The instance to update.
+            new_instance_status: The new status to transition to.
+            details: The details of the transition.
+
+        Raises:
+            ValueError if the transition is not allowed.
+        """
+        InstanceUtil._check_valid_next_instance_status(instance, new_instance_status)
+        instance.status = new_instance_status
+        InstanceUtil._record_status_transition(instance, new_instance_status, details)
+
+    @staticmethod
+    def _record_status_transition(
+        instance: Instance, status: Instance.InstanceStatus, details: str
+    ):
+        """Records the status transition.
+
+        This should be called by the instance manager.
+
+        Args:
+            instance: The instance to update.
+            status: The new status to transition to.
+        """
+        now_ms = time.time_ns() // 1000000
```

Comment:
Use the monotonically increasing timer.

Or record the wall clock time once at the beginning and then record elapsed time.

Reply:
Decided to just not enforcing the ordered timestamp since it's purely observability features.